packages/firestore/src/api/cache_config.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  IndexedDbOfflineComponentProvider,
+  MemoryOfflineComponentProvider,
+  MultiTabOfflineComponentProvider,
+  OfflineComponentProvider,
+  OnlineComponentProvider
+} from '../core/component_provider';
+
+export interface MemoryLocalCache {
+  kind: 'memory';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+}
+
+class MemoryLocalCacheImpl implements MemoryLocalCache {
+  kind: 'memory' = 'memory';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+
+  constructor() {
+    this._onlineComponentProvider = new OnlineComponentProvider();
+    this._offlineComponentProvider = new MemoryOfflineComponentProvider();
+  }
+}
+
+export interface IndexedDbLocalCache {
+  kind: 'indexeddb';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: OfflineComponentProvider;
+}
+
+class IndexedDbLocalCacheImpl implements IndexedDbLocalCache {
+  kind: 'indexeddb' = 'indexeddb';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: OfflineComponentProvider;
+
+  constructor(settings: IndexedDbSettings | undefined) {
+    let tabManager: IndexedDbTabManager;
+    if (settings?.tabManager) {
+      settings.tabManager.initialize(settings);
```

Comment:
I haven't read through everything yet, but flagging this to revisit. It's a little strange that this constructor initializes the `settings.tabManager` with the `settings` argument. This seems like it falls outside the responsibility of this class. It might feel better if this block checked if the tabManager was not already initialized.

Reply:
I agree with you. I could not come up with a better way however, because the way the underlying component providers need `cacheSizeBytes`, which comes from `setting` along size `tabManager`. Was hoping I can use your help to come up with a better way.